src/lib.rs

```diff
@@ -924,26 +932,17 @@ impl Drop for Rav1dContext {
         unsafe {
             if !(self.tc).is_null() {
                 let ttd: &TaskThreadData = &*self.task_thread;
-                if ttd.inited != 0 {
+                if self.n_tc > 1 {
                     let task_thread_lock = ttd.delayed_fg.lock().unwrap();
-                    let mut n: c_uint = 0 as c_int as c_uint;
-                    while n < self.n_tc
-                        && (*(self.tc).offset(n as isize)).task_thread.td.inited != 0
-                    {
-                        (*(self.tc).offset(n as isize)).task_thread.die = true;
-                        n = n.wrapping_add(1);
+                    for tc in &self.tc_shared {
+                        tc.thread_data.die.store(true, Ordering::Relaxed);
                     }
                     ttd.cond.notify_all();
                     drop(task_thread_lock);
-                    let mut n_0: c_uint = 0 as c_int as c_uint;
-                    while n_0 < self.n_tc {
-                        let pf: *mut Rav1dTaskContext =
-                            &mut *(self.tc).offset(n_0 as isize) as *mut Rav1dTaskContext;
-                        if (*pf).task_thread.td.inited == 0 {
-                            break;
+                    for tc in &self.tc_shared {
+                        if let Some(thread_id) = tc.handle {
+                            pthread_join(thread_id, 0 as *mut *mut c_void);
```

Comment:
Can you name `thread_id` `handle` instead?  Or something like `thread`.  Thread IDs are something else.